File: app-chip8-desktop/src/main/kotlin/fr/outadoc/kemu/chip8/desktop/Chip8DisplayView.kt
package fr.outadoc.kemu.chip8.desktop

import fr.outadoc.kemu.array.UByteArray2
import fr.outadoc.kemu.b
import fr.outadoc.kemu.chip8.Chip8Constants.DISPLAY_HEIGHT
import fr.outadoc.kemu.chip8.Chip8Constants.DISPLAY_WIDTH
import fr.outadoc.kemu.display.FrameConsumer
import fr.outadoc.kemu.theme.Theme
import fr.outadoc.kemu.theme.toColor
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Graphics2D
import javax.swing.JComponent

class Chip8DisplayView : FrameConsumer, JComponent() {

    private val scaleFactor = 10
    private var currentFrame: UByteArray2? = null

    var theme: Theme = Theme.WHITE_ON_BLACK
        set(value) {
            field = value
            background = value.background.toColor()
            foreground = value.foreground.toColor()
            repaint()
        }

    init {
        background = theme.background.toColor()
        foreground = theme.foreground.toColor()

        preferredSize = Dimension(
            (DISPLAY_WIDTH * scaleFactor),
            (DISPLAY_HEIGHT * scaleFactor)
        )
    }

    override fun displayFrame(frame: UByteArray2?) {
        currentFrame = frame
        repaint()
    }

    override fun paintComponent(g: Graphics?) {
        if (g !is Graphics2D) return

        if (currentFrame == null) {
            super.paintComponent(g)
            return
        }

        currentFrame?.forEachIndexed { i, pixel ->
            // Byte i in buffer is at screen position (i mod width, i / width)
            val x = i % DISPLAY_WIDTH
            val y = i / DISPLAY_WIDTH

            // Draw foreground color if bit is set, background otherwise
            g.color = if (pixel == 0x0.b) background else foreground
            g.fillRect(
                (x * scaleFactor), (y * scaleFactor), scaleFactor, scaleFactor
            )
        }
    }
}

File: app-chip8-ios/CHIP-8/Chip8DisplayView.swift
//
//  Chip8DisplayView.swift
//  CHIP-8 (iOS)
//
//  Created by Baptiste on 30/11/2020.
//

import Foundation
import UIKit
import lib_kemu
import lib_chip8

class Chip8DisplayView : UIView, FrameConsumer {
    
    private let scaleFactor = 10
    private var currentFrame: UByteArray2? = nil

    private var background: Color
    private var foreground: Color

    var theme: Theme = .whiteOnBlack {
        didSet {
            background = theme.background
            foreground = theme.foreground
            setNeedsDisplay()
        }
    }

    public required init?(coder aDecoder: NSCoder) {
        background = theme.background
        foreground = theme.foreground
        super.init(coder: aDecoder)
    }

    public override init(frame: CGRect) {
        background = theme.background
        foreground = theme.foreground
        super.init(frame: frame)
    }
    
    override var intrinsicContentSize: CGSize {
        return CGSize(
            width: (Int(Chip8Constants().DISPLAY_WIDTH) * scaleFactor),
            height: (Int(Chip8Constants().DISPLAY_HEIGHT) * scaleFactor)
        )
    }

    func displayFrame(frame: UByteArray2?) {
        currentFrame = frame
        setNeedsDisplay()
    }
    
    override func draw(_ rect: CGRect) {
        let ctx = UIGraphicsGetCurrentContext()
        if let frame = currentFrame {
            var i = 0
            while (frame.iterator().hasNext()) {
                let pixel = frame.iterator().next() as! NSNumber

                // Byte i in buffer is at screen position (i mod width, i / width)
                let x = i % Int(Chip8Constants().DISPLAY_WIDTH)
                let y = i / Int(Chip8Constants().DISPLAY_WIDTH)

                // Draw foreground color if bit is set, background otherwise
                let color = pixel == 0x0 ? background : foreground
                ctx?.setFillColor(color.toCGColor())

                ctx?.fill(
                    CGRect(
                        x: CGFloat(x * scaleFactor),
                        y: CGFloat(y * scaleFactor),
                        width: CGFloat(scaleFactor),
                        height: CGFloat(scaleFactor)
                    )
                )
                
                i += 1
            }
        } else {
            super.draw(rect)
        }
    }
}

extension Color {
    
    func toCGColor() -> CGColor {
        return CGColor(red: CGFloat(r), green: CGFloat(g), blue: CGFloat(b), alpha: 1.0)
    }
}
